src/main/java/net/fabricmc/loom/providers/MappingsProvider.java

```diff
@@ -66,50 +112,133 @@ public void provide(DependencyInfo dependency, Project project, LoomGradleExtens
 		project.getLogger().lifecycle(":setting up mappings (" + dependency.getDependency().getName() + " " + dependency.getResolvedVersion() + ")");
 
 		String version = dependency.getResolvedVersion();
-		File mappingsJar = dependency.resolveFile().orElseThrow(() -> new RuntimeException("Could not find dependency " + dependency));
+		File mappingsJar = dependency.resolveFile().orElseThrow(() -> new RuntimeException("Could not find yarn mappings: " + dependency));
+
+		// Hacky but seems like the easiest way to get the classifier
+		this.isV2 = mappingsJar.getName().endsWith("-v2.jar");
 
-		this.mappingsName = dependency.getDependency().getGroup() + "." + dependency.getDependency().getName();
+		this.mappingsName = StringUtils.removeSuffix(dependency.getDependency().getGroup() + "." + dependency.getDependency().getName(), "-unmerged");
+		if (this.isV2) mappingsName += "-v2";
 
 		Version mappingsVersion = new Version(version);
 		this.minecraftVersion = mappingsVersion.getMinecraftVersion();
 		this.mappingsVersion = mappingsVersion.getMappingsVersion();
 
 		initFiles(project);
 
-		if (!MAPPINGS_DIR.exists()) {
-			MAPPINGS_DIR.mkdir();
+		Files.createDirectories(mappingsDir);
+		Files.createDirectories(mappingsStepsDir);
+
+		if (!tinyMappings.exists()) {
+			storeMappings(project, minecraftProvider, mappingsJar);
 		}
 
-		if (!MAPPINGS_TINY_BASE.exists() || !MAPPINGS_TINY.exists()) {
-			if (!MAPPINGS_TINY_BASE.exists()) {
-				project.getLogger().lifecycle(":extracting " + mappingsJar.getName());
+		mappedProvider = new MinecraftMappedProvider();
+		mappedProvider.initFiles(project, minecraftProvider, this);
+		mappedProvider.provide(dependency, project, extension, postPopulationScheduler);
+	}
 
-				try (FileSystem fileSystem = FileSystems.newFileSystem(mappingsJar.toPath(), null)) {
-					Path fileToExtract = fileSystem.getPath("mappings/mappings.tiny");
-					Files.copy(fileToExtract, MAPPINGS_TINY_BASE.toPath());
-				}
-			}
+	private void storeMappings(Project project, MinecraftProvider minecraftProvider, File yarnJar) throws IOException {
+		if (isV2) {
+			// These are unmerged v2 mappings
+			Dependency intermediaryDep = project.getDependencies().add("compile", String.format("net.fabricmc:intermediary:%s:v2", minecraftProvider.minecraftVersion));
+			DependencyInfo intermediaryDepInfo = DependencyInfo.create(project, intermediaryDep, project.getConfigurations().getByName("compile"));
+			File intermediaryJar = intermediaryDepInfo.resolveFile().orElseThrow(() -> new RuntimeException("Could not find intermediaries: " + intermediaryDepInfo));
```

Comment:
Using the `compile` config here doesnt seem right? 

I dont think this would reslove the whole config but it may.

Reply:
Got a better idea?

Reply:
Honesly, here possibly download the file directally, or possibly create its own configuration?